highfidelity/hifi
# First step of cleaning up PointerEvents

We're cleaning up the tangled mess of mouse and pointer events in order to improve UI interaction.

This PR:
- simplifies the signals/slots we were using to move PointerEvents around
- events to web overlays/entities go directly to the target, instead of being emitted to all web overlays or web entities
- get rid of extraneous (and incorrectly-spelled) "doubleclickOnEntity" entity script callback
- fixes [FB2548](https://highfidelity.fogbugz.com/f/cases/2548/clickReleaseOnEntity-does-not-trigger-if-mouse-if-moved-off-instead-of-released)
- script-triggered PointerEvents properly trigger entity script event callbacks

Test Plan:
- Create several entities, including a couple web entities.  Also create a web overlay (you can use [this script](https://gist.githubusercontent.com/SamGondelman/a86b7d5f0a88929d73e3a3a06c1f3feb/raw/2fd2a04a553321e42d52a8f64d530e4ff81e37b3/UITest2.js) to make a big one).
**Run each of these tests in desktop mode with the mouse and also in HMD with the hand lasers**
- When you click on a web overlay or web entity, it should get focus (a yellow border will appear).  When you click off of them or on another object, they should lose focus (the yellow border will disappear) (this won't happen with the lasers, right now).  If a window has focus, it will receive key press events.  If you press space after clicking on a web entity, you should scroll down.  It should also work on a web overlay (try the marketplace tablet app).
- Click/drag on a web entity.  You should be able to select things and scroll.
- Click/drag on a web overlay.  You should be able to select things and scroll.  (It should also work on the tablet, but the behavior will likely be buggy like in master because of some separate issues)
- Context Overlays should still work:
    - Spawn something from the marketplace.
    - Right click on it (in HMD: hover with blue laser on it for 0.5 seconds).  The context overlay should appear (yellow outline, "i" circle overlay).
    - Hover over the "i".  It should appear solid.
    - Stop hovering over it.  Its transparency will animate.
    - Click on the "i".  The marketplace should open up.  Close it.
    - Right click on the entity again.  Left click on another entity.  The context overlay should disappear.
- Run [this](https://gist.githubusercontent.com/SamGondelman/f620a659cfd04db71101e292c079334d/raw/13ace804254090968ccf5a54cdbddb34b99ce0ed/EntityEventTest.js).  Click on an entity and, while clicking, drag off of it.  Release.  You should see a bunch of events logged, and the last one should be "clickReleaseOnEntity".  Stop that script.
- Run [this](https://gist.githubusercontent.com/SamGondelman/5e9fafa796693444a877b69ed388fb3d/raw/6e5071d04688fee60ee62c1babaacc80ae94baf9/EntityEventHandlerTest.js).  A box will appear in front of you.  Attach [this entity script to it](https://gist.githubusercontent.com/SamGondelman/34da6d658b9ccaf939abd7261b6c54aa/raw/fcc384b556e050d47803477e156263969c04b25b/EntityScriptEventTest.js).  Press "1".  You should see two events log: one from the local script, and one from the entity script (on master, you'd only see the local script).
- Repeatedly open and close the tablet while using the laser on it.  You should not crash.
- Switch domains while moving your mouse.  You shouldn't crash.
